system/jlib/jstring.cpp

```diff
@@ -935,22 +935,28 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
 }
 
 // Copy source to result, replacing all occurrences of "oldStr" with "newStr"
-StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char *source, size_t lenOldStr, const char* oldStr, size_t lenNewStr, const char* newStr)
+StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char *source, size_t lenOldStr, const char* oldStr, size_t lenNewStr, const char* newStr, bool & foundMatch)
 {
     if (lenOldStr && lenSource >= lenOldStr)
     {
-        // Avoid allocating an unnecessarly large buffer and match the source string
-        result.ensureCapacity(lenSource);
-
         size_t offset = 0;
         size_t lastCopied = 0;
         size_t maxOffset = lenSource - lenOldStr + 1;
         char firstChar = oldStr[0];
+        foundMatch = false;
         while (offset < maxOffset)
         {
             if (unlikely(source[offset] == firstChar)
                 && unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
             {
+                // Wait to allocate memory until a match is found
+                if (unlikely(!foundMatch))
```

Comment:
@ghalliday I am not sure I got this correct. I chose unlikely because it will only ever be true on the first match, so any string with multiple matches will get some benefit. If this is rare and in most cases there is only a single match then it should probably be likely. Is that correct?

Reply:
Most of the time I wouldn't worry about adding likely/unlikely.  The case where it is may be important is in inner loops that are the critical points.  Once we have a match it is much less critical.